report/common.py

```diff
@@ -0,0 +1,455 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Common libraries for report generation."""
+
+import dataclasses
+import io
+import json
+import logging
+import os
+import re
+from typing import List, Optional
+
+import yaml
+from google.cloud import storage
+
+import run_one_experiment
+from experiment import evaluator
+from experiment.workdir import WorkDirs
+
+MAX_RUN_LOGS_LEN = 16 * 1024
+
+TARGET_EXTS = ('.c', '.cc', '.cpp', '.cxx', '.c++', '.java', '.py')
+
+
+@dataclasses.dataclass
+class Benchmark:
+  """The class of a benchmark function and its experiment results."""
+  id: str
+  status: str
+  result: run_one_experiment.AggregatedResult
+  signature: str = ''
+  project: str = ''
+  function: str = ''
+
+
+@dataclasses.dataclass
+class Sample:
+  """Result of a fuzz target sample of a benchmark."""
+  id: str
+  status: str
+  result: evaluator.Result
+
+  @property
+  def stacktrace(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return f'{reproducer_link}/stacktrace'
+
+  @property
+  def target_binary(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return f'{reproducer_link}/target_binary'
+
+  @property
+  def reproducer(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return f'{reproducer_link}/artifacts'
+
+  @property
+  def run_log(self) -> str:
+    if not self.result:
+      return ''
+    reproducer_link = self.result.reproducer_path
+    return reproducer_link.removesuffix('reproducer') + 'run.log'
+
+
+@dataclasses.dataclass
+class Target:
+  code: str
+  fixer_prompt: Optional[str] = None
+
+
+class FileSystem:
+  """
+  FileSystem provides a wrapper over standard library and GCS client and
+  automatically chooses which to use based on the provided path.
+  """
+
+  _gcs_client = None
+
+  @classmethod
+  def _get_gcs_client(cls):
+    """
+    Returns a cached storage client (a new one is created on first call.
+
+    A new client does authentication on first call, so caching the client will
+    same multiple authentication round trips to GCP.
+    """
+    if cls._gcs_client is None:
+      cls._gcs_client = storage.Client()
+
+    return cls._gcs_client
+
+  def __init__(self, path: str):
+    logging.debug('file operation %s', path)
+    self._path = path
+    self._gcs_bucket: Optional[storage.Bucket] = None
+
+    if path.startswith('gs://'):
+      path = path.removeprefix('gs://')
+      self._gcs_bucket = FileSystem._get_gcs_client().bucket(path.split('/')[0])
+      self._path = '/'.join(path.split('/')[1:])
+
+  def listdir(self) -> List[str]:
+    """listdir returns a list of files and directories in path."""
```

Comment:
BTW, I noticed that all doc-string starts the function name, is this auto-generated or based on some convention?
Thanks!

Reply:
It's just a convention I like. Golang enforces function/module comments to start with the name which IMO generally makes golang docs more clear because it forces the sentences to focus on what the function/module does.